lib/spack/llnl/util/filesystem.py

```diff
@@ -782,22 +780,14 @@ def replace_directory_transaction(directory_name):
     directory_name = os.path.abspath(directory_name)
     assert os.path.isdir(directory_name), 'Not a directory: ' + directory_name
 
-    # Try to rename
-    for _ in range(10):
-        suffix = '_' + ''.join(random.choice(string.ascii_lowercase) for _ in range(5))
-        # normalized path does not have trailing dir separators
-        backup_dir = directory_name + suffix
-        try:
-            os.rename(directory_name, backup_dir)
-            last_error = None
-            break
-        except OSError as e:
-            last_error = e
-            continue
-
-    if last_error is not None:
-        raise last_error
+    # Note: directory_name is normalized here, meaning the trailing slash is dropped,
+    # so dirname is the directory's parent not the directory itself.
+    tmpdir = tempfile.mkdtemp(dir=os.path.dirname(directory_name))
 
+    # We have to jump through hoops to support Windows, since
+    # os.rename(directory_name, tmpdir) errors there.
+    backup_dir = os.path.join(tmpdir, 'backup')
+    os.rename(directory_name, backup_dir)
```

Comment:
Hm yeah, I kinda like rename since it guarantees the operation is fast or fails. I think it can only fail in some edge cases where prefixes are bind mounted from another filesystem.

Reply:
From the [`shutil`](https://docs.python.org/3/library/shutil.html#shutil.move) docs:

> If the destination is on the current filesystem, then [os.rename()](https://docs.python.org/2/library/os.html#os.rename) is used.

This has been there since `2.7`, and you're guaranteed same filesystem if you `mkdirp` right alongside the existing directory (unless some crazy person mounted an install prefix I guess).  So all it's doing is making it so you don't have to handle the ugly path details yourself.

But I'm fine with either.  If you still like `rename` I think the PR is good.